client/packages/programs/src/JsonForms/components/AdherenceStatus.tsx

```diff
@@ -0,0 +1,172 @@
+  const { data, handleChange, label, uischema, path } = props;
+  const [targetPath, setTargetPath] = useState<string | undefined>();
+  const [adherenceStatus, setAdherenceStatus] = useState<number | undefined>();
+
+  const { errors, options } = useZodOptionsValidation(
+    Options,
+    uischema.options
+  );
+
+  useEffect(() => {
+    if (!options) {
+      return;
+    }
+    const targetPath = composePaths(path, options.targetField);
+    setTargetPath(targetPath);
+    setAdherenceStatus(extractProperty(data, targetPath));
+  }, [options, path]);
+  useEffect(() => {}, [targetPath]);
+
+  // fetch current encounter
+  const encounterId = useEncounter.utils.idFromUrl();
+  const { data: currentEncounter, mutate: fetchEncounter } =
+    useEncounter.document.byId(encounterId);
+  useEffect(() => {
+    fetchEncounter();
+  }, [encounterId]);
+
+  // fetch previous encounter
+  const { data: previousEncounter, mutate: fetchPreviousEncounter } =
+    useEncounter.document.previous(
+      currentEncounter?.patient.id ?? '',
+      currentEncounter?.startDatetime
+        ? new Date(currentEncounter?.startDatetime)
+        : new Date()
+    );
+  useEffect(() => {
+    if (currentEncounter) {
+      fetchPreviousEncounter();
+    }
+  }, [currentEncounter]);
+
+  const params:
+    | {
+        previousCountOnHand: number;
+        remainingCount: number;
+        countPerDay: number;
+        timeDiffMs: number;
+      }
+    | undefined = useMemo(() => {
+    if (!options || !currentEncounter || !previousEncounter || !targetPath) {
+      return undefined;
+    }
+
+    const previousCountOnHand = extractProperty(
+      previousEncounter.document.data,
+      options.previousCountField
+    );
+    const remainingCount = extractProperty(data, options.remainingCountField);
+    const timeDiffMs =
+      new Date(currentEncounter.startDatetime).getTime() -
+      new Date(previousEncounter.startDatetime).getTime();
+
+    // Target pill count needed for the whole timespan from last till current encounter
+    const timeDiffDays = timeDiffMs / 1000 / 60 / 60 / 24;
+    const targetPillCount = timeDiffDays * options.countPerDay;
+    const status =
+      ((previousCountOnHand - remainingCount) / targetPillCount) * 100;
+
+    if (Number.isFinite(status) && status !== adherenceStatus) {
+      handleChange(targetPath, status);
+      setAdherenceStatus(status);
+    }
+
+    return {
+      previousCountOnHand,
+      remainingCount,
+      countPerDay: options.countPerDay,
+      timeDiffMs,
+    };
+  }, [options, previousEncounter, currentEncounter, data, targetPath]);
+
+  useEffect(() => {
+    if (!params || targetPath === undefined) {
+      return;
+    }
+  }, [params, targetPath]);
+
+  if (!props.visible) {
+    return null;
+  }
+
+  const inputProps = {
+    InputProps: {
+      sx: { width: '90px', '& .MuiInput-input': { textAlign: 'right' } },
+    },
+    disabled: true,
+    error: !!errors,
+    helperText: errors,
```

Comment:
Was thinking we should probably also show a specific error state for illegal (negative) values. If the user enters a higher "Remaining Pill Count" than there are available, then the adherence value is negative. It's probably too much hassle to have an additional check on the "Remaing pill count" field in order to prevent illegal values (would have to do a lookup to previous encounter to define its maximum), so might be easier just to show something informative (in red) in this component instead. (perhaps "N/A"?)

Reply:
Yes good point, not sure about it, maybe some warning? lets see if it could be a real issue I will check in meeting today